der/src/reader.rs

```diff
@@ -23,8 +23,10 @@ pub trait Reader<'r>: Sized {
     /// Get the length of the input.
     fn input_len(&self) -> Length;
 
-    /// Peek at the next byte of input without modifying the cursor.
-    fn peek_byte(&self) -> Option<u8>;
+    /// Peek at the byte of the input at position `offset` relative to the current position.
+    ///
+    /// Does not modify the decoder's state.
+    fn peek_offset(&self, offset: usize) -> Option<u8>;
```

Comment:
Changes to the reader trait like this will delay acceptance of this PR until we have a solution to #1228, FYI

Reply:
Tried to look around for alternatives. In reality I only need the `peek_tag` function now, peeking a byte at a specific offset is not necessarily needed.

Initially I thought, I could just do the same thing `SliceReader::peek_header()` is doing for the function `Reader::peek_tag()` (meaning clone and mutating read), but `peek_tag` is a provided method of `Reader`, which is not `Clone`.

That leaves two options as far as I can tell:

1. I could misuse `Reader::peek_header()` as a building block of `Reader::peek_tag()`: peeking the header and returning the tag. This is absolutely the wrong way around, but could allow the functionality without a change to the trait. [^1]
2. If it takes a while before this is merged, we will use our forked version until then. That is no issue at all

What do you think?

[^1]: I still think having the `Option` in `peek_tag(&self) -> Result<Option<Tag>, Error>` is pretty useful and could benefit the trait.

Reply:
#1418 contains part of the changes I want to make to the `Reader` trait. It should be a simplification for your purposes.

Regarding this method, perhaps it could be named `peek_byte_at` or thereabouts for consistency with the other methods which are named after the type they're peeking at?

Reply:
The methods return different types:

- `peek_tag` returns a `Tag`
- `peek_header` returns a `Header`
- `peek_byte` returns `u8`

While your proposal allows more of these to be provided methods (something that can also be implemented in terms of `peek_offset`, and would be generally nice to have), it doesn't actually provide equivalent functionality.